async-nats/src/service/mod.rs

```diff
@@ -0,0 +1,414 @@
+pub struct Info {
+    pub name: String,
+    pub id: String,
+    pub description: Option<String>,
+    pub version: String,
+    pub subject: String,
+}
+
+/// Schema of requests and responses.
+/// Currently, it does not do anything except providing information.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Schema {
+    pub request: String,
+    pub response: String,
+}
+
+/// Endpoint definition.
+/// In this iteration, it's just a subject.
+#[derive(Clone, Debug)]
+pub struct Endpoint {
+    pub subject: String,
+}
+
+/// Configuration of the [Service].
+#[derive(Debug)]
+pub struct Config {
+    pub name: String,
+    pub description: Option<String>,
+    pub version: String,
+    pub schema: Option<Schema>,
+    pub endpoint: String,
+}
+
+/// Verbs that can be used to acquire information from the services.
+pub enum Verb {
+    Ping,
+    Status,
+    Info,
+    Schema,
+}
+
+impl Display for Verb {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Verb::Ping => write!(f, "PING"),
+            Verb::Status => write!(f, "STATS"),
+            Verb::Info => write!(f, "INFO"),
+            Verb::Schema => write!(f, "SCHEMA"),
+        }
+    }
+}
+
+/// Service instance.
+///
+/// # Examples
+///
+/// ```no_run
+/// # #[tokio::main]
+/// # async fn main() -> Result<(), async_nats::Error> {
+/// use futures::StreamExt;
+/// let client = async_nats::connect("demo.nats.io").await?;
+/// let mut service = async_nats::service::add(client, async_nats::service::Config {
+///     name: "generator".to_string(),
+///     version: "1.0.0".to_string(),
+///     endpoint: "events.>".to_string(),
+///     schema: None,
+///     description: None,
+/// }).await?;
+///
+/// if let Some(request) = service.next().await {
+///     request.respond(Ok("hello".into())).await?;
+/// }
+///
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug)]
+pub struct Service {
+    stats: Arc<Mutex<Stats>>,
+    pub info: Info,
+    client: Client,
+    requests: Subscriber,
+    handle: JoinHandle<Result<(), Error>>,
+}
+
+/// Adds a new instance of the service.
+/// Every service with the same name scales horizontally.
+/// Services have their API for retrieving [StatsResponse], [Info], [Schema] and [Ping].
+///
+/// # Examples
+///
+/// ```no_run
+/// # #[tokio::main]
+/// # async fn main() -> Result<(), async_nats::Error> {
+/// use futures::StreamExt;
+/// let client = async_nats::connect("demo.nats.io").await?;
+/// let mut service = async_nats::service::add(client, async_nats::service::Config {
+///     name: "generator".to_string(),
+///     version: "1.0.0".to_string(),
+///     endpoint: "events.>".to_string(),
+///     schema: None,
+///     description: None,
+/// }).await?;
+///
+/// if let Some(request) = service.next().await {
+///     request.respond(Ok("hello".into())).await?;
+/// }
+///
+/// # Ok(())
+/// # }
+/// ```
+pub async fn add(client: Client, config: Config) -> Result<Service, Error> {
```

Comment:
Extension trait and hang this off client instead? maybe a pair of create_service and add_service methods?

```suggestion
pub async fn add(client: Client, config: Config) -> Result<Service, Error> {
```

Reply:
Yeah, was thinking about it.

Why a pair though?

Create to create a `Service` instance and `add` for interacting with it?